File: packages/hmssdk_flutter/ios/Classes/Models/HMSTranscriptExtension.swift
//
//  HMSTranscriptExtension.swift
//  hmssdk_flutter
//
//  Created by Pushpam on 14/06/24.
//

import Foundation
import HMSSDK

class HMSTranscriptExtension{
    
    static func toDictionary(hmsTranscript: HMSTranscript) -> [String:Any?]{
        
        var args = [String:Any?]()
        
        args["start"] = hmsTranscript.start
        args["end"] = hmsTranscript.end
        args["transcript"] = hmsTranscript.transcript
        args["peer_id"] = hmsTranscript.peer.peerID
        args["peer_name"] = hmsTranscript.peer.name
        args["is_final"] = hmsTranscript.isFinal
        
        return args
    }
    
    static func getMapFromTranscriptionsStateList(transcriptionStates: [HMSTranscriptionState]?) -> [[String:Any?]]?{
        
        if let transcripts = transcriptionStates{
            var transcriptStates = [[String:Any?]]()
            
            transcripts.forEach{
                transcriptStates.append(getMapFromTranscriptionState(transcriptionState: $0))
            }
            return transcriptStates
        }else{
            return nil
        }
        
    }
    
    private static func getMapFromTranscriptionState(transcriptionState: HMSTranscriptionState) -> [String:Any?]{
        
        var args = [String:Any?]()
        
        if let startedAt = transcriptionState.startedAt{
            args["started_at"] = Int(startedAt.timeIntervalSince1970 * 1000)
        }
        
        if let stoppedAt = transcriptionState.stoppedAt{
            args["stopped_at"] = Int(stoppedAt.timeIntervalSince1970 * 1000)
        }
        
        if let updatedAt = transcriptionState.updatedAt{
            args["updated_at"] = Int(updatedAt.timeIntervalSince1970 * 1000)
        }
        
        args["state"] = getStringFromTranscriptionState(transcriptionState: transcriptionState.state)
        
        args["mode"] = transcriptionState.mode
        
        return args
    }
    
    private static func getStringFromTranscriptionState(transcriptionState: HMSTranscriptionStatus) -> String?{
        switch transcriptionState{
        case .failed:
            return "failed"
        case .started:
            return "started"
        case .stopped:
            return "stopped"
        case .starting:
            return "initialized"
        default:
            return nil
        }
    }
    
    
    
}

File: packages/hmssdk_flutter/android/src/main/kotlin/live/hms/hmssdk_flutter/HMSTranscriptExtension.kt
package live.hms.hmssdk_flutter

import live.hms.video.sdk.models.OnTranscriptionError
import live.hms.video.sdk.models.TranscriptionState
import live.hms.video.sdk.models.Transcriptions
import live.hms.video.sdk.models.TranscriptionsMode
import live.hms.video.sdk.transcripts.HmsTranscript

class HMSTranscriptExtension {

    companion object{

        fun toDictionary(hmsTranscript: HmsTranscript):HashMap<String,Any?>{

            val args = HashMap<String,Any?>()

            args["start"] = hmsTranscript.start
            args["end"] = hmsTranscript.end
            args["transcript"] = hmsTranscript.transcript
            args["peer_id"] = hmsTranscript.peerId
            args["peer_name"] = hmsTranscript.peer?.name
            args["is_final"] = hmsTranscript.isFinal
            return args
        }

        fun getMapFromTranscriptionsList(transcriptions : List<Transcriptions>): ArrayList<HashMap<String,Any?>>{

            val transcriptionList = ArrayList<HashMap<String,Any?>>()

            transcriptions.forEach {_transcription ->
                transcriptionList.add(getMapFromTranscription(_transcription))
            }

            return transcriptionList
        }

        private fun getMapFromTranscription(transcription: Transcriptions): HashMap<String,Any?>{

            val args = HashMap<String,Any?>()

            args["error"] = transcriptionErrorExtension(transcription.error)
            args["started_at"] = transcription.startedAt
            args["stopped_at"] = transcription.stoppedAt
            args["updated_at"] = transcription.updatedAt
            args["state"] = getStringFromTranscriptionState(transcription.state)
            args["mode"] = getStringFromTranscriptionMode(transcription.mode)

            return args
        }

        private fun transcriptionErrorExtension(onTranscriptionError: OnTranscriptionError?): HashMap<String,Any?>?{
            onTranscriptionError?.let {_transcriptionError ->

                val args = HashMap<String,Any?>()

                args["code"] = _transcriptionError.code
                args["message"] = _transcriptionError.message

                return args

            }?: run {
                return null
            }
        }

        private fun getStringFromTranscriptionState(transcriptionState: TranscriptionState?): String?{
            return when(transcriptionState){
                TranscriptionState.STARTED -> "started"
                TranscriptionState.STOPPED -> "stopped"
                TranscriptionState.INITIALIZED -> "initialized"
                TranscriptionState.FAILED -> "failed"
                else -> null
            }
        }

        fun getTranscriptionModeFromString(transcriptionMode: String): TranscriptionsMode?{
            return when(transcriptionMode){
                "caption" -> TranscriptionsMode.CAPTION
                "live" -> TranscriptionsMode.LIVE
                else -> null
            }
        }

        fun getStringFromTranscriptionMode(transcriptionMode: TranscriptionsMode?): String?{
            return when(transcriptionMode){
                TranscriptionsMode.CAPTION -> "caption"
                TranscriptionsMode.LIVE -> "live"
                else -> null
            }
        }
    }
}
